### PR TITLE
[trainer] fix: Loss calculations for grad accumulation steps

### PR DESCRIPTION
**Pull request by puneeshkhanna**
### What does this PR do?

For gradient accumulation steps over micro batches, loss should be normalised before calling loss.backwards(). 
Also add an optimisation so that the all reduce of gradients is only performed in the last accumulation step.

Verified fine-tuning on few open source models with the changes in this PR.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
@vermouth1992 - please review this. Grad accumulation loss calculation is incorrect in SFT code for micro batch steps. I think the loss calculation is correct in RL code for grad accumulation over micro batch steps.

**Comment by puneeshkhanna** [comment]
Addressed precommit reqs of line too long.

**Comment by Schilings** [comment]
Regarding the loss calculation during gradient accumulation: if the loss is computed using a token-level strategy where the number of valid tokens can vary across different micro-batches, wouldn't scaling the loss simply by 1 / num_micro_batches be inaccurate?
image

**Comment by puneeshkhanna** [comment]
@Schilings - What I fixed in this PR is loss averaging for gradient accumulation steps before calling loss.backwards(). In RL code as an example here - https://github.com/volcengine/verl/blob/a4d8952edc6653e9661bbe37722ede13f2a7b0c7/verl/workers/actor/dp_actor.py#L482 it is done like that and it is the standard way. 
Regarding the doubt of valid tokens per DP rank, I think one can leverage balance_dp_token flag.

**Comment by Schilings** [comment]
@puneeshkhanna Thank you for your reply. I understand that the work in your PR addresses a different issue, and I'm not raising this as a concern with your changes. I just wanted to discuss this specific point with you.

The total training batch size is calculated as micro_batch_size_per_gpu * dp_size * grad_acc_steps. I've looked at the source code, and as far as I understand, the balance_dp_token flag ensures that within each gradient accumulation step, different DP ranks can calculate a consistent total number of valid tokens across ranks. However, the number of valid tokens can still vary between different gradient accumulation steps.

Given this potential variation in valid tokens per step, wouldn't directly scaling the loss for each step simply by 1 / num_micro_batches be mathematically inequivalent? Is this level of approximation generally considered acceptable, or is there a mechanism I might be missing that accounts for the differing token counts per step during the final gradient update?

**Comment by Schilings** [comment]
the direct use of 1 / n_micro_batches for averaging the loss per step in gradient accumulation is indeed a common simplification. The main goal is to align the gradient scale with what you'd get from an equivalent large batch size, and it's definitely simpler to implement.

I now understand that while it might not be mathematically precise when the number of valid tokens varies per micro-batch (especially with token-level loss), it's generally considered an acceptable approximation in practice.

File: verl/trainer/fsdp_sft_trainer.py [hunk 1]
```diff
@@ -365,7 +365,10 @@ def _compute_loss_and_backward(self, batch, do_backward=True):
 
         # Context manager for sequence parallel if needed
         context = self.sharding_manager if use_sp else nullcontext()
-        with context, torch.autocast(device_type=self.device_name, dtype=torch.bfloat16):
+        # If last micro-batch in an accumulation step, then implicit grad all-reduce happens in loss backward.
+        # Otherwise, use no_sync() to avoid communication overhead during intermediate accumulation steps.
+        sync_context = nullcontext() if last_micro_in_accum else self.fsdp_model.no_sync()
```
Comment (vermouth1992) [inline verl/trainer/fsdp_sft_trainer.py]: We don't want to use no_sync here as it will cause higher memory consumption. To prevent silent errors, we disable CPUOffloading explicitly for FSDP1

Comment (puneeshkhanna) [reply verl/trainer/fsdp_sft_trainer.py]: Removed that part of PR

**Comment by puneeshkhanna** [comment]
@vermouth1992 - please check now updated PR

**Comment by vermouth1992** [comment]
Could you please fix pre-commit?

**Comment by puneeshkhanna** [comment]
> Could you please fix pre-commit?

@vermouth1992 - Done
